### PR TITLE
feat(builder): add .git to submodule builds

### LINKED ISSUE #3349: Submodules support without .git
Reading through #3305 and having had my PR #3172 closed, I'm left scratching my head here.

In the PR review for #3172 this morning, I'm sure you discussed how submodules would work with #3305. I'm just not seeing it.

Consider this issue a documentation request on how someone can make use of submodules without including the .git directory.

### PR DESCRIPTION
**Pull request by ianblenke**
feat(deis-builder/optional_submodules): Add .git to builds

Add .git to builds if .gitmodules is present to allow for submodules.

Closes #3349.

### CODE REVIEW

**Comment by bacongobbler** [comment]
Could you please amend your commit such that it follows the [commit style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style)? Thanks!

**Comment by ianblenke** [comment]
Git commit amended to follow the commit style guide.

**Comment by mboersma** [comment]
This code looks good. Sorry to ask again, but could you amend the commit message like this:
```
feat(builder): add .git to submodule builds
```
The idea is to keep the (scope) simple and usually matching a directory, and to use a lower-case verb. :mag:

**Comment by ianblenke** [comment]
Cleaned up the commit message. Thanks for the quick pointers, that helped. I'll make sure to follow that for future PRs.

**Comment by mboersma** [comment]
Code LGTM.

**Comment by mboersma** [comment]
In testing this, I provisioned a new vagrant cluster off master + this PR. I pushed a copy of our example-go app, then committed a trivial git submodule (from another public GitHub repo) to it and did `git push deis master` again. But I get this:
```console
$ git push deis master
Counting objects: 4, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 404 bytes | 0 bytes/s, done.
Total 3 (delta 1), reused 0 (delta 0)
Cloning into '/home/git/jovial-tailbone.git/build/tmp.du1jyBoB5C'...
fatal: '/home/git/jovial-tailbone.git/build' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
To ssh://git@deis.local3.deisapp.com:2222/jovial-tailbone.git
   40be416..127586d  master -> master
$ ls -al
total 56
drwxr-xr-x  10 matt  staff    340 Mar 31 16:02 .
drwx------  84 matt  staff   2856 Mar 31 15:32 ..
drwxr-xr-x  15 matt  staff    510 Mar 31 16:02 .git
-rw-r--r--   1 matt  staff     93 Mar 31 16:02 .gitmodules
drwxr-xr-x   5 matt  staff    170 Mar 31 15:32 Godeps
-rw-r--r--   1 matt  staff  10273 Mar 31 15:32 LICENSE
-rw-r--r--   1 matt  staff     16 Mar 31 15:32 Procfile
-rw-r--r--   1 matt  staff   2043 Mar 31 15:32 README.md
drwxr-xr-x   2 matt  staff     68 Mar 31 16:02 test-jargon
-rw-r--r--   1 matt  staff    660 Mar 31 15:32 web.go
```
@ianblenke am I testing this correctly, and did it work ok for you?

**Comment by ianblenke** [comment]
Looking at it now.

**Comment by ianblenke** [comment]
Not entirely sure how that actually worked for me before. After hammering at it briefly, this iteration does appear to work for me now.

**Comment by ianblenke** [comment]
Did an interactive rebase and a force push to normalize the commit message as per the commit style guide.

**Comment by ianblenke** [comment]
Learned a lesson... it can't be a shallow clone, otherwise it loses the history of most of the stuff in the folder - so much so that a 'git status' shows that the .gitmodules and other files are dirty and pending a git add for the HEAD state. So now this just does a full clone, and doesn't try any of the weirdness. Fully tested, at least I _know_ this works now.

File: builder/image/templates/builder [hunk 1]
```diff
@@ -63,8 +63,20 @@ mkdir -p $BUILD_DIR $CACHE_DIR
...
+git config core.bare false
+git reset --hard $GIT_SHA
+
+# If there is not a .gitmodules in this non-bare clone, revert to git archive behavior
+if [ ! -f $TMP_DIR/.gitmodules ]; then
+  TMP_DIR=$(mktemp -d --tmpdir=$BUILD_DIR)
+
+  cd $REPO_DIR
+
+  git archive $GIT_SHA | tar -xmC $TMP_DIR
```
Comment (bacongobbler) [inline builder/image/templates/builder]: At this point, TMP_DIR already contains the source code from the clone, so basically this is a no-op. We need to start from a fresh slate here or else `.git` will still be present after the `git archive`.

**Comment by bacongobbler** [comment]
Jenkins tends to disagree with you:

> Cloning into '/home/git/appssample.git/build/tmp.SlhLUjOux4'...
warning: You appear to have cloned an empty repository.
done.
fatal: Not a git repository: '.'
To ssh://git@deis.test-199dd08b.deis.works:2222/appssample.git

**Comment by mboersma** [comment]
@ianblenke this is still failing on a regular `git push` in the integration tests. Any idea why?

**Comment by mboersma** [comment]
Let's re-open this if/when we can figure out why it breaks `git push`.

**Pull request closed without merging**